### PR TITLE
[DONT_MERGE] Feature code quality (fix #4)

### LINKED ISSUE #4: State of the source code
At the moment, mics is advertise as

> Intuitive mixins for ES6 classes

However, in its own code, it uses a wild mix of outdated syntax, like `var` and not even available syntax (at least on NodeJS), like `export function`. I think, it would be better to choose one. Personally, I would prefer ES6 as is available on NodeJS LTS (v6 atm), as it has a lot of benefits, especially for static analysis, and gets rid of the dependency on a transpiler for deployment.

Additionally, there is no linting available, and the code seems to not follow any guidelines. E.g. sometimes there are multiple variables declared per `var` statement, sometimes there are multiple consecutive `var` statements. I would suggest using [airbnb's coding guidelines](https://github.com/airbnb/javascript) in order to improve the code quality and consistency.

After #2 is finished, I could run the code through WebStorm code analysis and fix the syntax one way or the other, if desired. It's only a few lines, so not too much trouble, actually :)

### PR DESCRIPTION
**Pull request by minecrawler**
code restructuring.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,75 +1,112 @@
...
-	return Object.defineProperties(constructor, {
-		mixin: {value:mixin, writable:false},
-		class: {value: Class, writable:false},
-		interface: {get:(x => () => x ? x : x = getInterface(Class.prototype))()},
-	})
-}
+const baseclass = class Object {};
+const derive = superclass => ({}[superclass.name || 'Object'] = class extends superclass {});
+
+const getPropertyNames = function getPropertyNames(proto) {
```
Comment (Download) [inline src/index.js:6]: What exactly is the reasoning behind this change? What does the `const getPropertyNames = ` bring us? I know functions are hoisted, but all this code is packaged in a wrapper function anyway.

File: src/index.js [hunk 2]
```diff
@@ -1,75 +1,112 @@
...
-		var c = x
-		while (c !== Object) {
-			if (c === type || c === type.class) return true
-			if (type.mixin && type.mixin.classes && type.mixin.classes.indexOf(c) !== -1) return true
-			c = c.prototype.__proto__.constructor
-		}
-	}
-	return false
-}
+    if (typeof type === 'string') {
```
Comment (Download) [inline src/index.js:29]: So I know ESLint makes us do this... change `==` to `===`. It's one of the reasons I don't use ESLint normally. However this adds bytes to my production build. Uglify and friends can't know it's actually safe to use `==` here. So what are we buying from these extra bytes?

Could I convince you to configure ESLint to be loose on the triple equals rule? I really *hate* that rule. 
Most of these ESLint changes I don't like, but I want this project to be usable by many teams so probably linting it is the way to go. But that triple equals rule... I hate it with passion.

File: src/index.js [hunk 3]
```diff
@@ -1,75 +1,112 @@
...
 }
 
-function getPropertyNames(proto) {
-	var results = []
-	while (proto !== Object.prototype) {
-		Object.getOwnPropertyNames(proto).reduce((arr,k) => arr.indexOf(k) === -1 ? arr.push(k) && arr : arr, results)
-		proto = proto.__proto__.constructor.prototype
-	}
-	return results
+function mix(...args) {
```
Comment (Download) [inline src/index.js:62]: Can we maybe put this back at the top?

In the original source code, the public API stuff was at the very top and the further down you go, the more private/implementation specific/low-level the code gets. That was actually done on purpose. That way someone casually inspecting the source does not have to wade through a whole lot of nitty gritty low level private functions before he gets to the ones he is used to using in his program.

**Review by Download — COMMENTED** [review]
Just wrote some comments/questions.

Thanks for helping out! I see we have some very different code styles ha ha! But I think you are absolutely right to implement linting. It would be good to convince others that we take code quality and consistency serious. However maybe I can plead with you to make the linting config a bit more loose ha ha.  :)

File: package.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "mics",
-  "version": "0.5.0",
+  "version": "0.6.1",
```
Comment (Download) [inline package.json:3]: I think it's best to just ignore versions in the feature branches. Once the feature branch is merged, we cut a new version directly from master. This prevents the feature branch having to guess at the version, or getting conflicts because the version changed from another feature being merged.

File: src/index.js (hunk 1 above)
Comment (minecrawler) [reply src/index.js:6]: I recommend reading the airbnb guide, on which the linting is based. It is quite good for modern projects and explains its choices well. https://github.com/airbnb/javascript#functions

File: src/index.js (hunk 2 above)
Comment (minecrawler) [reply src/index.js:29]: What you trade here is usually stability vs. size. A few extra bytes don't really matter, as we have internet speeds in the megabytes/second readily available in many regions and it is a lot more likely that the your user is sloppy on their end and you end up comparing a string to a number or similar.

However, I will change the triple-equal back in places which feel safe, if that's important to you.

File: src/index.js (hunk 3 above)
Comment (minecrawler) [reply src/index.js:62]: Well, you make use of JS function hoisting here, which in general is contra-intuitive. Logically, you first declare and define, then use something, however, putting the export and high-level functions to the top is the other way around.

I haven't seen a library using that kind of sorting, but if you want it that way, I can comply.

File: package.json (hunk 4 above)
Comment (minecrawler) [reply package.json:3]: Well, you might want to think about using [GitFlow](http://nvie.com/posts/a-successful-git-branching-model/), which is a lot more clear about that kind of thing. The version and final changes would be committed to a release branch, before landing in master, which always only contains a production-ready version.

File: src/index.js (hunk 1 above)
Comment (Download) [reply src/index.js:6]: I'm sorry, can you summarize? I see zero advantages atm.

File: src/index.js (hunk 3 above)
Comment (Download) [reply src/index.js:62]: > you make use of JS function hoisting here. 

Correct. I actually like that feature, because it allows me to put what's important at the top. In fact I think that's what hoisting was meant to achieve.

> I haven't seen a library using that kind of sorting

Actually it's pretty common in many languages. Pascal comes to mind. Otherwise you are forced to put the low level bits at the top.

File: package.json (hunk 4 above)
Comment (Download) [reply package.json:3]: Haven't used it before but I'm open to trying it 👍

File: src/index.js (hunk 2 above)
Comment (Download) [reply src/index.js:29]: > However, I will change the triple-equal back in places which feel safe, if that's important to you.

It would be much appreciated,

File: src/index.js (hunk 3 above)
Comment (Download) [reply src/index.js:62]: I've been thinking some more about this.

In a perfect world, `index.js` would contain just some imports. So import `mix.js`, `is.js` and `like.js` for example. And then export them under the common namespace. As a new coder, you open `index.js` and see where the stuff is coming form and then open the correct file. This gives the top-down overview of the code I want to achieve. But the problem is that actually splitting up this code into 4 files gives way too much overhead. Webpack modules are powerful but they come at a price. I found this from carefully examining the packaged code.

So this technique using function hoisting is a compromise where you still get a top-down view but I can cram everything into one file to reduce overhead. Pascal did the same thing where one `unit` contained many functions, classes etc.

File: src/index.js (hunk 1 above)
Comment (Download) [reply src/index.js:6]: Been reading the AirBnB guide but afaict their argument is just 'hoisting is bad'.
Personally I feel hoisting is a feature that is not that hard to understand and actually has some advantages as well. But hey I also love the ternary operator `?` and the ugly `|| defaultValue` trick. I like shortcuts. Because they are short ;) 

My problem with linting is that it ends up as a religious debate.

Ask yourself: why did AirBnB create their own coding standard? Are they really the first to do this? Why didn't they adopt e.g. https://standardjs.com/ or any of the dozens of other 'standards' in use?

Comment (Download) [reply src/index.js:6]: You might not believe it, but I actually think the original 75 lines of code were more easy to read then the new 112. That's probably a personal thing but I've noticed that I'm most productive if code is close together so as much as possible of it will fit on one screen. Having to scroll is actually hurting my productivity, which is why I will write `if (..) <statement>` on a single line and do more stuff like that that breaks convention.

**Comment by minecrawler** [comment]
> You might not believe it, but I actually think the original 75 lines of code were more easy to read then the new 112. That's probably a personal thing but I've noticed that I'm most productive if code is close together so as much as possible of it will fit on one screen. Having to scroll is actually hurting my productivity, which is why I will write if (..) <statement> on a single line and do more stuff like that that breaks convention.

Well. I found several instances of functions like this:

```js
var foo = (() => [1,2,3].reduce((s, v) => s + v))();
```

and this

```js
var bar = (() => x => [1,2,3].reduce((s, v) => s + v) + x)(baz);
```

That's over-engineered and unperformant and doesn't make sense. Additionally, instead of using a very simple switch-statement with a good overview, you concatenated 3 or 4 conditions with logical and ternary operators, leading to a mess I was not able to grasp at all until I carefully pulled it apart.

Opensource cuts both ways. You write source and you have to be productive, however, others who want to contribute also have to be able to read and understand the source. Since I was not able to understand the original source at all, others will feel the same way. On top of that, I have the impression that you yourself have problems understanding your own code, as you tend to over-engineer it a lot.

So, if you want to keep using your style, that's ok and I respect your decision, but I will not be able to contribute and I will not feel confident using your solution. However, if you think we can compromise, then I am onboard. I decided for the airbnb style, because I think it is safe, leads to easy to read code and is consistent with other C-like languages. If you do not like airbnb's style, we can use a different one, like JSSS.

**Comment by Download** [comment]
@minecrawler 
Sorry if I came across as rigid. I'm a kinda stubborn person with strong opinions. 

However I am dedicated to working as a team member on this and be open to input from you (and others willing to contribute). That said, I feel completely rewriting the code from scratch (as this PR is doing) is counter-productive.

How about this compromise:

You setup linting like in this PR, but instead of changing all the code so the linter doesn't compain, we start by making the linting (much) more loose so we have to change as little code as possible. For example I deliberately chose a semicolon-less style, so maybe we can change the linter to accomodate this.

Once this is merged, we can create a/some separate PR(s) to re-enable linting on those rules that we agree are valuable.

> Opensource cuts both ways. You write source and you have to be productive, however, others who want to contribute also have to be able to read and understand the source.

Totally agree! And I should start by admitting that I am biased towards my own style. So I value your input! Hopefully we can find a middle ground that makes us both happy.

> That's over-engineered and unperformant and doesn't make sense. 

I am assuming you are referring to e.g.
```js
if (args.length) factory = (org => superclass => org(args.reduce((s,m) => m.mixin(s), superclass)))(factory)
``` 
and
```js
interface: {get:(x => () => x ? x : x = getInterface(Class.prototype))()},
```
I admit I may have gone a bit overboard here.  :-)  However I do not agree with the underperfomant aspect of your comment. In fact performance was very much on my mind when I wrote this code. A big part of perf in JS comes from a *small* library, which is why this code is written so terse. And in the second example, the `x` is a 'private' variable that is used to cache the result of the `getInterface` call.

However 'permature optimization is the root of all evil' they always say, so probably we should refactor this to be more readable  👍 

**EDIT**: At the risk of making a long comment even longer, let me add this observation. This PR does a lot of stuff. Also, some stuff I see would be very useful to have, right away. But currently our 'debate' (lol) about coding style is standing in the way of a merge, 

May I suggest you create a couple of smaller PRs:
* [x] [Add Jetbrains project to ignored folders](https://github.com/Download/mics/issues/10)
* [ ] [Package lock for NPM](https://github.com/Download/mics/issues/11)
* [ ] [Setup loose linting](https://github.com/Download/mics/issues/12)
* [ ] [Create PR(s) for linting rules](https://github.com/Download/mics/issues/13)

I can then quickly merge them. 

I think once this is done, you should [become collaborator of this project](https://github.com/Download/mics/milestone/2), so you don't have to wait around for me to merge stuff.  👍

**Comment by minecrawler** [comment]
Thank you for your understanding. I guess it is important that we agree on the changes and make a compromise for the best future of the project :)

So, let's close this PR and make smaller ones. I will shift further discussion on the mentioned topics to the appropriate locations

**Comment by Download** [comment]
Great!
